src/lib.rs

```diff
@@ -94,3 +93,188 @@ mod tests {
+/// This prevents path traversal attacks.
+///
+/// It just allows a single normal path element and no parent, root directory or prefix like `C:`.
+#[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
+pub struct PathComponent {
+    path: PathBuf,
+}
+
+impl PathComponent {
+    /// It creates the wrapped `PathComponent` if it's valid.
+    /// Otherwise it will return `None`.
+    ///
+    /// ```
+    /// # use pathbuf::PathComponent;
+    /// # #[cfg(unix)]
+    /// # {
+    /// let some_valid_folder: PathComponent = PathComponent::new("foo").unwrap();
+    /// let some_valid_file: PathComponent = PathComponent::new("bar.txt").unwrap();
+    /// assert!(PathComponent::new("/etc/shadow").is_none());
+    /// # }
+    /// ```
+    pub fn new<S: Into<PathBuf>>(component: S) -> Option<Self> {
+        let component = Self {
+            path: component.into(),
+        };
+
+        component.is_valid().then_some(component)
+    }
+
+    #[cfg(feature = "sanitise")]
+    /// This will sanitise the input and therefore all inputs are valid.
+    /// Unless there is a bug in the sanitisation then it would `panic`.
+    ///
+    /// ```
+    /// # use pathbuf::PathComponent;
+    /// # #[cfg(unix)]
+    /// # {
+    /// assert_eq!(
+    ///     PathComponent::with_sanitise("/etc/shadow"),
+    ///     PathComponent::new("etc_shadow").unwrap(),
+    /// );
+    /// # }
+    /// ```
+    ///
+    /// The sanitisation algorithm isn't considered stable.
+    /// Therefore the sanitised path could change in the future for the same input.
+    pub fn with_sanitise(component: &str) -> Self {
+        let sanitised_component = sanitize_filename_reader_friendly::sanitize(component);
+        Self::new(sanitised_component).unwrap_or_else(|| {
+            panic!(
+                "Expected a sanitised path of the original path '{}'",
+                component
+            )
+        })
+    }
+
+    fn is_valid(&self) -> bool {
+        use std::path::Component;
+
+        let mut components = self.path.components();
+        matches!(
+            (components.next(), components.next()),
+            (Some(Component::Normal(_)), None)
+        )
+    }
+}
+
+impl std::ops::Deref for PathComponent {
+    type Target = std::path::Path;
+
+    fn deref(&self) -> &Self::Target {
+        &self.path
+    }
+}
+
+impl AsRef<std::path::Path> for PathComponent {
+    fn as_ref(&self) -> &std::path::Path {
+        &self.path
+    }
+}
+
+/// This allows to push just a [`PathComponent`] to a [`std::path::PathBuf`].
+///
+/// ```
+/// use std::path::PathBuf;
+/// # use pathbuf::{pathbuf, PathComponent, PushPathComponent};
+/// # #[cfg(unix)]
+/// # {
+/// let mut path = PathBuf::new();
+/// path.push_component(PathComponent::new("foo").unwrap());
+/// path.push_component(PathComponent::new("bar.txt").unwrap());
+///
+/// assert_eq!(path, pathbuf!["foo", "bar.txt"])
+/// # }
+/// ```
+pub trait PushPathComponent {
```

Comment:
Let's delete the extension trait. It's not saving much.

Reply:
I think we have different ideas about the features of this PR. I wanted a solution on the type-level, which could also be integrated into the macro.
Using the lower-level types has the following benefits in some situations:
- flexibility: multi-path segments can be allowed in any position
- [Parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/): user input can be converted early in the program and functions can take the `SinglePathComponent`

But I can slim the PR down if these features are not wanted.